instrumentation/cassandra/src/main/java/brave/cassandra/Tracing.java

```diff
@@ -0,0 +1,149 @@
+
+/**
+ * This creates Zipkin server spans for incoming cassandra requests. Spans are created when there's
+ * a tracing component available, and the incoming payload is not explicitly unsampled.
+ *
+ * <h3>Configuring a tracing component</h3>
+ * If the system property "zipkin.http_endpoint" is set, a basic tracing component is setup.
+ *
+ * <p>Otherwise, {@link brave.Tracing#current()} is used. This relies on external bootstrapping of
+ * {@link brave.Tracing}.
+ *
+ * <p>Alternatively, you can subclass this and fix configuration to your favorite mechanism.
+ */
+public class Tracing extends org.apache.cassandra.tracing.Tracing {
+  final InetAddress coordinator = FBUtilities.getLocalAddress();
+  final TracingComponent component;
+
+  public Tracing(brave.Tracing tracing) { // subclassable to pin configuration
+    component = new TracingComponent.Explicit(tracing);
+  }
+
+  public Tracing() {
+    String endpoint = System.getProperty("zipkin.http_endpoint");
+    if (endpoint == null) {
+      component = new TracingComponent.Current();
+      return;
+    }
+    brave.Tracing tracing = brave.Tracing.newBuilder()
+        .localServiceName(System.getProperty("zipkin.service_name", "cassandra"))
+        .reporter(AsyncReporter.create(URLConnectionSender.create(endpoint)))
+        .build();
+    component = new TracingComponent.Explicit(tracing);
+  }
+
+  /**
+   * When tracing is enabled and available, this tries to extract trace keys from the custom
+   * payload. If that's possible, it re-uses the trace identifiers and starts a server span.
+   * Otherwise, a new trace is created.
+   */
+  @Override protected final UUID newSession(UUID sessionId, TraceType traceType,
+      Map<String, ByteBuffer> customPayload) {
+    Tracer tracer = component.tracer();
+    if (tracer == null || traceType == TraceType.NONE) {
+      return super.newSession(sessionId, traceType, customPayload);
+    }
+    Span span = spanFromPayload(tracer, customPayload).kind(SERVER);
+
+    // override instead of call from super as otherwise we cannot store a reference to the span
+    assert get() == null;
+    TraceState state = new ZipkinTraceState(coordinator, sessionId, traceType, span);
+    set(state);
+    sessions.put(sessionId, state);
+    return sessionId;
+  }
+
+  /** This extracts the RPC span encoded in the custom payload, or starts a new trace */
+  Span spanFromPayload(Tracer tracer, Map<String, ByteBuffer> payload) {
+    TraceContextOrSamplingFlags contextOrFlags = payload == null
+        ? TraceContextOrSamplingFlags.create(SamplingFlags.EMPTY)
+        : component.extractor().extract(payload);
+    return contextOrFlags.context() != null
+        ? tracer.joinSpan(contextOrFlags.context())
+        : tracer.newTrace(contextOrFlags.samplingFlags());
+  }
+
+  @Override protected final void stopSessionImpl() {
+    ZipkinTraceState state = (ZipkinTraceState) get();
+    if (state != null) state.incoming.finish();
+  }
+
```

Comment:
how come you didn't also override `public void doneWithLocalSession(TraceState)` ?